src/goto-programs/string_refine_preprocess.cpp

```diff
@@ -0,0 +1,1343 @@
+  {
+    args.push_back(function_call.arguments()[i]);
+    new_type.parameters().push_back(function_type.parameters()[i]);
+  }
+
+  exprt lhs;
+  if(assign_first_arg)
+    lhs=function_call.arguments()[0];
+  else
+    lhs=function_call.lhs();
+
+  if(lhs.id()==ID_typecast)
+    lhs=to_typecast_expr(lhs).op();
+
+  new_type.return_type()=lhs.type();
+
+  make_string_function(
+    goto_program, i_it, lhs, new_type, function_name, args, loc, signature);
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::make_string_function_call
+
+  Inputs: a position in a goto program and a function name
+
+ Purpose: at the current position, replace `s.some_function(x,...)` by
+          `s=function_name(x,...)`
+
+\*******************************************************************/
+
+void string_refine_preprocesst::make_string_function_call(
+  goto_programt &goto_program,
+  goto_programt::targett &i_it,
+  const irep_idt &function_name,
+  const std::string &signature)
+{
+  make_string_function(
+    goto_program, i_it, function_name, signature, true, true);
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::make_string_function_side_effect
+
+  Inputs: a position in a goto program and a function name
+
+ Purpose: at the current position, replace `r=s.some_function(x,...)`
+          by `s=function_name(s,x,...)` and add a correspondance from r
+          to s in the `string_builders` map
+
+\*******************************************************************/
+
+void string_refine_preprocesst::make_string_function_side_effect(
+  goto_programt &goto_program,
+  goto_programt::targett &i_it,
+  const irep_idt &function_name,
+  const std::string &signature)
+{
+  const code_function_callt &function_call=to_code_function_call(i_it->code);
+  string_builders[function_call.lhs()]=function_call.arguments()[0];
+  make_string_function(
+    goto_program, i_it, function_name, signature, true, false);
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::build_function_application
+
+  Inputs: a function name, a type, a location and a vector of arguments
+
+ Outputs: a function application expression
+
+ Purpose: declare a function and construct an function application expression
+          with the given function name, type, location and arguments
+
+\*******************************************************************/
+
+function_application_exprt
+  string_refine_preprocesst::build_function_application(
+    const irep_idt &function_name,
+    const typet &type,
+    const source_locationt &location,
+    const exprt::operandst &arguments)
+{
+  declare_function(function_name, type);
+  function_application_exprt function_app;
+  function_app.type()=type;
+  function_app.add_source_location()=location;
+  function_app.function()=symbol_exprt(function_name);
```

Comment:
There should really be a two-argument constructor for function_application_exprt (consuming a symbol_exprt and a type). Not your fault, but it would be great if you could fix that and use it here!